### ISSUE TITLE
Streaming member selection leads to inconsistent results

### ISSUE THREAD

**Issue by nivekastoreth** [stedolan/jq#2464]
**Describe the bug**
Attempting to build sub-objects from streaming data by filtering the keys leads to inconsistent results with simple changes to the underlying data structure

**To Reproduce**

Attempting to build an object with the members filtered based on streaming data results in no object being created due to the lack of an object-close stream element
```
$ jq -nc '{http:["data"],tcp:["data"]}' | jq -nc --stream 'inputs'
[["http",0],"data"]
[["http",0]]
[["tcp",0],"data"]
[["tcp",0]]
[["tcp"]]

$ jq -nc '{http:["data"],tcp:["data"]}' | jq -nc --stream 'fromstream(inputs|select(.[0][0] == "http"))' # produces no output
```

If the same is done where the object is already filtered, it produces output as expected
```
$ jq -nc '{http:["data"]}' | jq -nc --stream 'inputs'
[["http",0],"data"]
[["http",0]]
[["http"]]

$ jq -nc '{http:["data"]}' | jq -nc --stream 'fromstream(inputs|select(.[0][0] == "http"))'
{"http":["data"]}
```

**Expected behavior**
Filtering streaming objects by key should behave the same regardless of the underlying data
```
$ jq -nc '{http:["data"],tcp:["data"]}' | jq -nc --stream 'fromstream(inputs|select(.[0][0] == "http"))'
{"http":["data"]}
```

**Environment (please complete the following information):**
```
$ jq --version
jq-1.6

$ uname -a
Linux jmaki-pc 5.10.60.1-microsoft-standard-WSL2 #1 SMP Wed Aug 25 23:20:18 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

$ lsb_release -d
Description:    Ubuntu 20.04.4 LTS
```

**Additional context**
N/A

**Comment by pkoppstein** [comment]
It's not clear to me that this is actually a bug in jq or your program.  In any case, `fromstream` expects an "end of object" marker that your filter excludes.  I believe the simplest generic workaround is to ensure there is a suitable marker by providing [["null"]]:

```
echo '{"http":["data"], "tcp":["data"]}' |
  jq -nc --stream 'fromstream( (inputs|select(.[0][0] == "http")), [[null]])' 
```

**Comment by nivekastoreth** [comment]
Given the new `pick` feature, I could rephrase this as a feature request for a streaming version of `pick`, since what I'm looking for is the ability to pick the elements I want from a stream without having to reconstitute the objects first
